diff -Nur subversion-1.9.4-orig/configure subversion-1.9.4/configure
--- subversion-1.9.4-orig/configure	2016-04-21 18:55:50.000000000 +0400
+++ subversion-1.9.4/configure	2016-09-05 20:06:35.359215141 +0400
@@ -19259,7 +19259,7 @@
   if test "$cross_compiling" = yes; then :
   { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
 $as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
-as_fn_error $? "cannot run test program while cross compiling
+$as_echo $? "cannot run test program while cross compiling
 See \`config.log' for more details" "$LINENO" 5; }
 else
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
